src/libraries/Microsoft.Extensions.DependencyInjection/src/ServiceLookup/ServiceProviderEngineScope.cs

```diff
@@ -155,30 +161,53 @@ static async ValueTask Await(int i, ValueTask vt, List<object> toDispose)
                         ((IDisposable)disposable).Dispose();
                     }
                 }
+
+                scope.ClearState();
+            }
+        }
+
+        private IDictionary<ServiceCacheKey, object> ScopeDisposed()
+        {
+            ThrowHelper.ThrowObjectDisposedException();
+            return null;
+        }
+
+        private void ClearState()
+        {
+            // We lock here since ResolvedServices is always accessed in the scope lock, this means we'll never
+            // try to return to the pool while somebody is trying to access ResolvedServices.
+            lock (_scopeLock)
```

Comment:
this method seems to be a noop for root scope, would it make sense to skip for it before taking the lock?